### ISSUE TITLE
Mathematical Reasoning in Neural Networks

### ISSUE THREAD

**Issue by howardyclo** [howardyclo/papernotes#52]
### My Reading List (To be done ^_^)
My research question: We know that deep learning can learn to "solve" math if given training data like (e.g, x+y=z, x-y=z, ...) with x, y, z in a **specific range**. BUT, what if x, y, z becomes **out-of the range** of training data? (I think it will failed ..., which is also a problem of naive supervised learning). Maybe the paper below will help me figure that out.

### Representative people:
- [Felix Hill (DeepMind)](https://scholar.google.co.uk/citations?hl=en&user=4HLUnhIAAAAJ&view_op=list_works&sortby=pubdate)

### Papers
- [Analysing  Mathematical Reasoning Abilities of Neural Models](https://arxiv.org/pdf/1904.01557.pdf?fbclid=IwAR0v0M8WFeYrXygkOmpmRQ2xwBCNoAtHra5ox5RjswoqR4c69imPzRgGb88). by David Saxton et al. ICLR 2019. DeepMind.
- [Neural Arithmetic Logic Units](https://arxiv.org/pdf/1808.00508.pdf) by Andrew Trask et al. DeepMind.

**Comment by howardyclo** [comment]
## Neural Arithmetic Logic Units
### DNNs with Non-linearities Struggle to Learn Identity Function
![](https://github.com/howardyclo/papernotes/blob/master/images/NALU/2.png?raw=true)
- Train an autoencoder to reconstruct its input ranged [-5, 5].
- All autoencoders are identical in its parameterization (3 hidden layers of size 8), only using different non linearities.
- Trained with MSE loss. 
- Tested in [-20, 20], the error increase severely both below and above the range of numbers seen during training.

### The Neural Accumulator (NAC) & Neural Arithmetic Logit Unit (NALU)
![](https://github.com/howardyclo/papernotes/blob/master/images/NALU/1.png?raw=true)
- NAC: A special case of linear layer, whose weight matrix W only consists of {-1, 0, 1}, defined as:
    - W = tanh(\hat{W}) * σ(\hat{M})
    - The elements of W are guaranteed to be [-1, 1], and biased towards {-1, 0, 1} during learning, since {-1, 0, 1} corresponds to the saturation points of either tanh(.) or σ(.)
    - Its output are additions or subtractions of rows in the input vector.
- NALU: Learns a weighted sum between two sub-cells:
    - One is the original NAC,  capable of learning to add and subtract.
    - The other one operates in log space, capable of multiply and divid, e.g., log(XY) = logX + logY; log(X/Y) = logX - log Y; exp(log(X)) = X
